src/vm_pmu.rs

```diff
@@ -368,4 +373,149 @@ impl VmPmuState {
             }
         }
     }
+
+    /// Helper function to call either SBI, or the internal implementation
+    /// for configure_matching_counters() depending on whether Sscdeleg is implemented
+    pub fn configure_matching_counters(
+        &self,
+        counter_index: u64,
+        counter_mask: u64,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<u64> {
+        if !CpuInfo::get().has_sscdeleg() {
+            sbi::api::pmu::configure_matching_counters(
+                counter_index,
+                counter_mask,
+                config_flags,
+                event_type,
+                event_data,
+            )
+        } else {
+            fn is_firmware_counter(event_type: PmuEventType) -> bool {
+                matches!(event_type, PmuEventType::Firmware(_))
+            }
+
+            fn is_fixed_counter(counter_index: usize) -> bool {
+                // CYCLE and INSRET are fixed counters
+                counter_index == 0 || counter_index == 2
+            }
+
+            fn is_timer(counter_index: u64) -> bool {
+                counter_index == 1
+            }
+
+            if is_firmware_counter(event_type) {
+                return Err(SbiError::InvalidParam);
+            }
+
+            let pmu_info = pmu::PmuInfo::get()?;
+            let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+            let index = if !config_flags.is_skip_match() {
+                let mut bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
+                bitmask_iter
+                    .find(|i| {
+                        matches!(self.counter_state[*i], NotConfigured) || is_fixed_counter(*i)
+                    })
```

Comment:
Hm, is this match condition right? As written, we'd always select a fixed counter index if it were included in the counter mask, even if the requested event wasn't a fixed counter event. Shouldn't we only be picking a fixed counter if the user requested their respective events? e.g. always pick counter 0 if user requested the `cycles` event.

Reply:
Since the iterator filters elements based on the counter_mask, find() should iterate only over those elements, meaning that we'll hit the condition only if a fixed counter was selected by both the counter_index and counter_mask, right?

```
// Convenience iterator to return counter_state indexes relative to counter_index
// if the counter_mask bit is set (assumes sanitized input).
impl Iterator for CounterMaskIter {
    type Item = usize;
    fn next(&mut self) -> Option<Self::Item> {
        loop {
            if self.counter_mask == 0 {
                break None;
            }
            let result = self.counter_index + self.mask_index;
            let mask_bit_set = self.counter_mask & (1 << self.mask_index);
            self.counter_mask &= !(1 << self.mask_index);
            self.mask_index += 1;
            if mask_bit_set != 0 {
                break Some(result as usize);
            }
        }
    }
}
```

Reply:
Yes, but I'm suggesting that we should only be selecting a fixed counter if the requested event type matches.

Reply:
Ah, definitely makes sense.

Reply:
Changed...given the conversation about the preferred approach, there's no need for us to treat fixed counters differently. If the fixed counter is selected, and is available, it will be automatically picked since it'll be a lower index. if not, we'll simply choose one of the available configurable counters, and the assumption is that the platform can use it for fixed counters as well.

Reply:
Changed to ensure static mapping for fixed counters per offline chat with @atishp04